frame/benchmarking/src/lib.rs

```diff
@@ -695,6 +799,110 @@ macro_rules! benchmark_backend {
 	};
 }
 
+// Creates #[test] functions for the given bench cases.
+#[macro_export]
+#[doc(hidden)]
+macro_rules! impl_bench_case_tests {
+	(
+		{ $module:ident, $new_test_exec:expr, $exec_name:ident, $test:path, $extra:expr }
+		{ $( $names_extra:tt )* }
+		$( { $( $bench_inst:ident )? } $bench:ident )*
+	)
+	=> {
+		$crate::impl_bench_name_tests!(
+			$module, $new_test_exec, $exec_name, $test, $extra,
+			{ $( $names_extra )* },
+			$( { $bench } )+
+		);
+	}
+}
+
+// Creates a #[test] function for the given bench name.
+#[macro_export]
+#[doc(hidden)]
+macro_rules! impl_bench_name_tests {
+	// recursion anchor
+	(
+		$module:ident, $new_test_exec:expr, $exec_name:ident, $test:path, $extra:expr,
+		{ $( $names_extra:tt )* },
+		{ $name:ident }
+	) => {
+		$crate::paste::paste! {
+			#[test]
+			fn [<bench_ $name>] () {
+				$new_test_exec.$exec_name(|| {
+					// Skip all #[extra] benchmarks if $extra is false.
+					if !($extra) {
+						let disabled = $crate::vec![ $( stringify!($names_extra).as_ref() ),* ];
+						if disabled.contains(&stringify!($name)) {
+							$crate::log::error!(
+								"INFO: extra benchmark skipped - {}",
+								stringify!($name),
+							);
+							return ();
+						}
+					}
+
+					// Same per-case logic as when all cases are run in the
+					// same function.
+					match std::panic::catch_unwind(|| {
+						$module::<$test>::[< test_benchmark_ $name >] ()
+					}) {
+						Err(err) => {
+							println!(
+								"{}: {:?}",
+								stringify!($name),
+								err,
+							);
+							assert!(false);
+						},
+						Ok(Err(err)) => {
+							match err {
+								$crate::BenchmarkError::Stop(err) => {
+									println!(
+										"{}: {:?}",
+										stringify!($name),
+										err,
+									);
+									assert!(false);
+								},
+								$crate::BenchmarkError::Override(_) => {
+									// This is still considered a success condition.
+									$crate::log::error!(
+										"WARNING: benchmark error overrided - {}",
+										stringify!($name),
+									);
+								},
+								$crate::BenchmarkError::Skip => {
+									// This is considered a success condition.
+									$crate::log::error!(
+										"WARNING: benchmark error skipped - {}",
+										stringify!($name),
+									);
+								}
+							}
+						},
+						Ok(Ok(())) => (),
+					}
```

Comment:
I don't think we need to do this complex stuff, and I think this complex gives not good error message, which was the point of the PR. Instead we should just makes panic being regular panic in test, and we should give the error when the benchmark return an error.
So something like this
```suggestion
					// Same per-case logic as when all cases are run in the
					// same function.
					match $module::<$test>::[< test_benchmark_ $name >] () {
						Err(err) => {
							match err {
								$crate::BenchmarkError::Stop(err) => panic!("benchmark return stop: {}", err),
								// This is still considered a success condition.
								$crate::BenchmarkError::Override(_) => (),
								// This is still considered a success condition.
								$crate::BenchmarkError::Skip => (),
							}
						},
						Ok(()) => (),
					}
```

Reply:
also here I don't see why we log a warning where it is a correct implementation of a benchmark.

Reply:
The problem here is that you may want to program in a a custom `BenchmarkError` for using the benchmark in production, but not want to trigger that pathway when running a test.

The problem is that not all benchmarks apply to all possible runtime configurations (specifically in the XCM world), which is why i wanted to funnel this data to the top

Reply:
Ok my suggestion is wrong, we can keep the log. But moving the println into the panic message gives probably a better error message.

Reply:
I felt a bit weird for writing that `assert!(false)` :laughing:.
Replaced it with and the `println` with `panic!("{}: {:?}", stringify!($name), err);`.